lib/stateManager.js

```diff
@@ -249,14 +440,39 @@ proto.dumpStorage = function (address, cb) {
   })
 }
 
+/**
+ * Callback for `hasGenesisState` method
+ * @callback hasGenesisState~callback
+ * @param {Error} error an error that may have happened or `null`
+ * @param {Boolean} hasGenesisState Whether the storage trie contains the
+ * canonical genesis state for the configured chain parameters.
+ */
+
+/**
+ * Checks whether the current instance has the canonical genesis state
+ * for the configured chain parameters.
+ * @memberof DefaultStateManager
+ * @method hasGenesisState
+ * @param {hasGenesisState~callback} cb
+ */
 proto.hasGenesisState = function (cb) {
   const root = this._common.genesis().stateRoot
   this._trie.checkRoot(root, cb)
 }
 
+/**
+ * Generates a canonical genesis state on the instance based on the
+ * configured chain parameters. Will error if there are uncommitted
+ * checkpoints on the instance.
+ * @memberof StateManager
+ * @method generateCanonicalGenesis
+ * @param {Function} cb Callback function
+ */
 proto.generateCanonicalGenesis = function (cb) {
   var self = this
 
+  if (self._checkpointCount !== 0) { return cb(new Error('Cannot create genesis state with uncommitted checkpoints')) }
+
   this.hasGenesisState(function (err, genesis) {
     if (!genesis && !err) {
       self.generateGenesis(genesisStates[self._common.chainName()], cb)
```

Comment:
Here there is a call from a method from the interface to a method from the default implementation, this is probably not intended?

Reply:
If I understand you right it's completely intended. The method `generateCanonicalGenesis` in the implementation can use whatever mechanism it likes to provide implementation. This is much like how `putContractStorage` uses `_modifyContractStorage` internally. 

FYI, I did consider removing `generateCanonicalGenesis` from the interface completely, but it would require further changes as to how we initialize the state trie. Depending on feedback during the beta this is something we might change still.

Reply:
We would extract both to a single repo yes. Presuming it would be Typescript, we would then define the interface along with a reference (default) implementation in the external repo with both types being exported. Now there is a question as to whether the reference implementation should include methods beyond the interface? My gut instinct is it should not and we should strip away the extra functionality. It just felt like too much to do as part of this refactor